File: LeetCodePractice/DataStructure/Heap/Heap.py
class MinHeap:
	'''MinHeap class for numeric list'''
	def __init__(self, arr = []):
		self.heap = arr
	
	#get parent index from childIndex
	def parent(self, childIndex):
		if(childIndex == 0):
			return 0
		return (childIndex+1)//2-1

	#get left child index from parentIndex
	def left(self, parentIndex):
		return (parentIndex*2)+1

	#get right child index from parentIndex
	def right(self, parentIndex):
		return (parentIndex*2)+2

	#move element at index i down to proper place so heap properity is maintained
	def heapifyDown(self, i):
		left, right, smallest = self.left(i), self.right(i), i
		if(left < len(self.heap) and self.heap[left] < self.heap[i]):
			smallest = left
		if(right < len(self.heap) and self.heap[right] < self.heap[smallest]):
			smallest = right
		if(smallest != i):
			self.swap(smallest, i)
			self.heapifyDown(smallest)

	#move element at index i up to proper place so heap properity is maintained
	def heapifyUp(self, i):
		parent = self.parent(i)
		if(i != parent and self.heap[i] < self.heap[parent]):
			self.swap(i, parent)
			self.heapifyUp(parent)

	#swap values of elements at heap index x and y
	def swap(self, x, y):
		self.heap[x], self.heap[y] = self.heap[y], self.heap[x]

	#heapity an unordered array, assuming arr is integer array
	def heapifyArray(self, arr):
		self.heap = arr
		i = len(self.heap)//2
		while(i >= 0):
			self.heapifyDown(i)
			i -= 1

	#insert an element to heap, and maintain heap property
	def insert(self, element):
		self.heap.append(element)
		self.heapifyUp(len(self.heap)-1)

	#return the minimum element in the heap without removing it
	def peek(self):
		if(len(self.heap) > 0):
			return self.heap[0]
		else:
			return None

	#delete an element in index i from heap
	def deleteAt(self, i):
		if(i != len(self.heap)-1):
			self.swap(i, len(self.heap)-1)
			del self.heap[len(self.heap)-1]
			self.heapifyDown(i)
		else:
			del self.heap[i]

	#delete an element in heap, return  -1 if not found
	def deleteEle(self, element):
		elementIndex = self.heap.index(element)
		self.deleteAt(elementIndex)

	#return the minimum element in the heap and remove it
	def pop(self):
		if(len(self.heap) > 0):
			result = self.heap[0]
			self.deleteAt(0)
			return result
		else:
			return None

	#print the heap
	def __str__(self):
		s = ''
		for ele in self.heap:
			s += str(ele) + " "
		return s

#some sample test
heap = MinHeap()

heap.insert(286789035)
heap.insert(255653921)
heap.insert(274310529)
heap.insert(494521015)
print(heap.peek())
heap.deleteEle(255653921)
heap.deleteEle(286789035)
print(heap.peek())
heap.insert(236295092)
heap.insert(254828111)
heap.deleteEle(254828111)
heap.insert(85886315)
heap.insert(7959587)
heap.insert(20842598)
heap.insert(494521015)
heap.deleteEle(7959587)		
print(heap.peek())

File: LeetCodePractice/DataStructure/Heap/Heap.cs
using System.Collections.Generic;

namespace LeetCodePractice.DataStructure.Heap
{
    public class Heap<T>
    {
        private readonly IList<T> _data;
        private readonly IComparer<T> _comparer;
        public int Count => _data.Count;
        public T Peek => _data[0];        

        public Heap(IList<T> data, IComparer<T> comparer = null)
        {
            _data = data;
            _comparer = comparer ?? Comparer<T>.Default;
            //heapify data
            int i = _data.Count / 2;
            while(i >= 0)
            {
                Down(i--);
            }
        }

        //push a node down
        private void Down(int i)
        {
            int left = Left(i), right = Right(i), smallest = i;
            if (left < _data.Count && _comparer.Compare(_data[left], _data[i]) < 0) smallest = left;
            if(right < _data.Count && _comparer.Compare(_data[right], _data[smallest]) < 0) smallest = right;
            if(smallest != i)
            {
                Swap(smallest, i);
                Down(smallest);
            }
        }

        //bubble up a node
        private void Up(int i)
        {
            int p = Parent(i);
            if(i != p && _comparer.Compare(_data[p], _data[i]) > 0)
            {
                Swap(p, i);
                Up(p);
            }
        }

        //return the first value, and maintain heap
        public T Pop()
        {
            if(_data.Count > 0)
            {
                T res = _data[0];
                DeleteAt(0);
                return res;
            }
            return default(T);
        }

        //add an element
        public void push(T t)
        {
            _data.Add(t);
            Up(_data.Count - 1);
        }
        
        //delete an element at index i
        public void DeleteAt(int i)
        {
            if(i != _data.Count - 1)
            {
                Swap(i, _data.Count - 1);
                _data.RemoveAt(_data.Count - 1);
                Down(i);
            }
            else
            {
                _data.RemoveAt(i);
            }
        }

        private int Parent(int child)
        {
            return child == 0 ? 0 : (child + 1) / 2 - 1;
        }
        private int Left(int parent)
        {
            return parent * 2 + 1;
        }
        private int Right(int parent)
        {
            return parent * 2 + 2;
        }

        private void Swap(int i, int j)
        {
            T temp = _data[i];
            _data[i] = _data[j];
            _data[j] = temp;
        }
    }
}
